### PR TITLE
Task manager shard repair tasks

### PR DESCRIPTION
**Pull request by Deexie**
The PR introduces shard_repair_task_impl which represents a repair task
that spans over a single shard repair.

repair_info is replaced with shard_repair_task_impl, since both serve 
similar purpose.

### CODE REVIEW

**Comment by Deexie** [comment]
- rebased

**Comment by Deexie** [comment]
@bhalevy review please

File: repair/repair_task.hh [hunk 1]
```diff
@@ -72,6 +72,26 @@ protected:
...
+    shard_repair_task_impl(tasks::task_manager::module_ptr module,
+            tasks::task_id id,
+            std::string keyspace,
+            std::string type,
+            tasks::task_id parent_id,
+            lw_shared_ptr<repair_info> ri,
+            std::exception_ptr ex)
+        : repair_task_impl(module, id, 0, std::move(keyspace), "", std::move(type), "", parent_id)
+        , _ri(ri)
+        , _ex(ex)
```
Comment (bhalevy) [inline repair/repair_task.hh]: nit: std::move(ex)

File: repair/repair.cc [hunk 2]
```diff
@@ -549,102 +549,112 @@ get_sharder_for_tables(seastar::sharded<replica::database>& db, const sstring& k
...
+    repair_uniq_id parent_id_,
     const std::vector<sstring>& data_centers_,
     const std::vector<sstring>& hosts_,
     const std::unordered_set<gms::inet_address>& ignore_nodes_,
     streaming::stream_reason reason_,
+    abort_source* as,
     bool hints_batchlog_flushed)
-    : rs(repair)
+    : repair_task_impl(module, id, 0, keyspace, "", std::move(type), "", parent_id_.uuid())
+    , _ex(ex)
```
Comment (bhalevy) [inline repair/repair.cc:569]: nit: std::move(ex)

**Review by bhalevy — COMMENTED** [review]
nice transition!

**Comment by asias** [comment]
Looks good. Thanks.

@Deexie Per my previous ask, could you add some examples how the repair task api would look like?

**Comment by Deexie** [comment]
> Looks good. Thanks.
> 
> @Deexie Per my previous ask, could you add some examples how the repair task api would look like?

@asias I am not sure what you mean. Would you like me to paste a request that is necessary to get the repair related data or to describe the request parameters or maybe something else?

Also what is the right place for this information? Currently, we can check the repair task status, but there is no meaningful information about task progress, since this is going to be added together with per range tasks. So maybe the api examples should be put in the cover-letter of per range repair PR when the api provides complete task status?

**Comment by Deexie** [comment]
- added std::move to shard_repair_task_impl constructor

**Comment by bhalevy** [comment]
> > Looks good. Thanks.
> > @Deexie Per my previous ask, could you add some examples how the repair task api would look like?
> 
> @asias I am not sure what you mean. Would you like me to paste a request that is necessary to get the repair related data or to describe the request parameters or maybe something else?
> 
> Also what is the right place for this information? Currently, we can check the repair task status, but there is no meaningful information about task progress, since this is going to be added together with per range tasks. So maybe the api examples should be put in the cover-letter of per range repair PR when the api provides complete task status?

Let me add that the repair api is basically kept changed.
The task manager api provides extra functionality on top it by allowing to list tasks, get their progress, and abort them (if supported by the task) in an orderly and standard way.

File: repair/repair.cc [hunk 3]
```diff
@@ -455,8 +458,11 @@ bool repair_module::is_aborted(const tasks::task_id& uuid) {
 void repair_module::abort_all_repairs() {
     _aborted_pending_repairs = _pending_repairs;
     for (auto& x : _repairs) {
-        auto& ri = x.second;
-        ri->abort();
+        auto it = _tasks.find(x.second);
+        if (it != _tasks.end()) {
+            auto& impl = dynamic_cast<shard_repair_task_impl&>(*it->second->_impl);
+            impl.abort_repair_info();
```
Comment (bhalevy) [inline repair/repair.cc:464]: I guess we can't use a generic repair method for this yet
but this is addressed in a follow up series.
If so, please add a respective FIXME comment

**Review by bhalevy — COMMENTED** [review]
LGTM other than the FIXME comment around the need for integration with the generic task abort functionality

**Comment by asias** [comment]
> > > Looks good. Thanks.
> > > @Deexie Per my previous ask, could you add some examples how the repair task api would look like?
> > 
> > 
> > @asias I am not sure what you mean. Would you like me to paste a request that is necessary to get the repair related data or to describe the request parameters or maybe something else?
> > Also what is the right place for this information? Currently, we can check the repair task status, but there is no meaningful information about task progress, since this is going to be added together with per range tasks. So maybe the api examples should be put in the cover-letter of per range repair PR when the api provides complete task status?
> 
> Let me add that the repair api is basically kept changed. The task manager api provides extra functionality on top it by allowing to list tasks, get their progress, and abort them (if supported by the task) in an orderly and standard way.

I meant concrete example outpust of the those functions like list tasks / get progress.

$ curl htttp://127.0.0.1/list_tasks

$  curl  htttp://127.0.0.1/get_task_progress

**Comment by Deexie** [comment]
@asias Here is the example output of the api calls (on 3 node cluster). There are two more calls: wait_task and abort_task. 
The first has the same output as task_status, though it waits until the task is finished (failed or done).
The latter does not make sense yet, since aborting task is out of scope of this PR.

task_manager/list_modules:
```
[
  "repair"
]
```

task_manager/list_module_tasks/repair: (can be filtered by keyspace/table)
```
[
  {
    "task_id": "9aa65703-5152-426e-8138-501b48339fe3",
    "state": "done"
  },
  {
    "task_id": "5314013f-e169-4b18-8326-cab3f83088d9",
    "state": "done"
  },
  {
    "task_id": "9bb2cd68-246b-4241-8fd6-e1784516ef51",
    "state": "done"
  },
  {
    "task_id": "d901a6c0-7784-4b88-98c7-f9bb5ab9fee2",
    "state": "done"
  },
  {
    "task_id": "bc8ae5ef-c252-4d10-9650-3eed359f8cc9",
    "state": "done"
  },
  {
    "task_id": "6d424f3e-51b7-4a69-aaed-58052e744880",
    "state": "done"
  },
  {
    "task_id": "241a9d1c-0ef6-467e-93b2-1508349db2ec",
    "state": "done"
  },
  {
    "task_id": "ac674e45-2f17-445b-9407-33ee5d91ed16",
    "state": "done"
  }
]
```

task_manager/task_status/9bb2cd68-246b-4241-8fd6-e1784516ef51:
```
{
  "id": "9bb2cd68-246b-4241-8fd6-e1784516ef51",
  "type": "repair",
  "state": "done",
  "is_abortable": false,
  "start_time": "2022-11-30T11:52:31Z",
  "end_time": "2022-11-30T11:52:31Z",
  "error": "",
  "parent_id": "9aa65703-5152-426e-8138-501b48339fe3",
  "sequence_number": 4,
  "shard": 0,
  "keyspace": "ks",
  "table": "",
  "entity": "",
  "progress_units": "ranges",
  "progress_total": 0,
  "progress_completed": 0
}
```

**Review by denesb — APPROVED** [review]

**Comment by Deexie** [comment]
- added FIXME comment

**Comment by bhalevy** [comment]
> task_manager/list_module_tasks/repair: (can be filtered by keyspace/table)

@Deexie for humans  I think we need more information in the tasks list like the task type and a bit of metadata like keyspace, table, and entity, (maybe shard too but I don't remember if those are listed by default) so one can choose which task to drill into to get its particular status or to abort it.  Otherwise you'll need to the status of each tasks to know what they are about.

**Comment by Deexie** [comment]
> > task_manager/list_module_tasks/repair: (can be filtered by keyspace/table)
> 
> @Deexie for humans I think we need more information in the tasks list like the task type and a bit of metadata like keyspace, table, and entity, (maybe shard too but I don't remember if those are listed by default) so one can choose which task to drill into to get its particular status or to abort it. Otherwise you'll need to the status of each tasks to know what they are about.

I will submit the change to api in separate PR, since its quite unrelated to this one, ok?

**Comment by bhalevy** [comment]
> > > task_manager/list_module_tasks/repair: (can be filtered by keyspace/table)
> > 
> > 
> > @Deexie for humans I think we need more information in the tasks list like the task type and a bit of metadata like keyspace, table, and entity, (maybe shard too but I don't remember if those are listed by default) so one can choose which task to drill into to get its particular status or to abort it. Otherwise you'll need to the status of each tasks to know what they are about.
> 
> I will submit the change to api in separate PR, since its quite unrelated to this one, ok?

of course

**Comment by asias** [comment]
> @asias Here is the example output of the api calls (on 3 node cluster). There are two more calls: wait_task and abort_task. The first has the same output as task_status, though it waits until the task is finished (failed or done). The latter does not make sense yet, since aborting task is out of scope of this PR.
> 
> task_manager/list_modules:
> 
> ```
> [
>   "repair"
> ]
> ```
> 
> task_manager/list_module_tasks/repair: (can be filtered by keyspace/table)

Thanks. It looks like the tasks listed here contains not only the higher level task but also the child tasks. How many levels do we actually have?  E.g., the one created by user from api (highest level), then each keysapces / tables / ranges, each shard?

> ```
> [
>   {
>     "task_id": "9aa65703-5152-426e-8138-501b48339fe3",
>     "state": "done"

How long will the "done" tasks stay? When will we get rid of the tasks that have finished? 

>   },
>   {
>     "task_id": "5314013f-e169-4b18-8326-cab3f83088d9",
>     "state": "done"
>   },
>   {
>     "task_id": "9bb2cd68-246b-4241-8fd6-e1784516ef51",
>     "state": "done"
>   },
>   {
>     "task_id": "d901a6c0-7784-4b88-98c7-f9bb5ab9fee2",
>     "state": "done"
>   },
>   {
>     "task_id": "bc8ae5ef-c252-4d10-9650-3eed359f8cc9",
>     "state": "done"
>   },
>   {
>     "task_id": "6d424f3e-51b7-4a69-aaed-58052e744880",
>     "state": "done"
>   },
>   {
>     "task_id": "241a9d1c-0ef6-467e-93b2-1508349db2ec",
>     "state": "done"
>   },
>   {
>     "task_id": "ac674e45-2f17-445b-9407-33ee5d91ed16",
>     "state": "done"
>   }
> ]
> ```
> 
> task_manager/task_status/9bb2cd68-246b-4241-8fd6-e1784516ef51:
> 
> ```
> {
>   "id": "9bb2cd68-246b-4241-8fd6-e1784516ef51",
>   "type": "repair",
>   "state": "done",
>   "is_abortable": false,
>   "start_time": "2022-11-30T11:52:31Z",
>   "end_time": "2022-11-30T11:52:31Z",
>   "error": "",
>   "parent_id": "9aa65703-5152-426e-8138-501b48339fe3",
>   "sequence_number": 4,

What is the sequence number? How is it numbered?

>   "shard": 0,
>   "keyspace": "ks",
>   "table": "",
>   "entity": "",

What is entity?

>   "progress_units": "ranges",
>   "progress_total": 0,
>   "progress_completed": 0

Does the progress actually work now?

> }
> ```

How do I know if the task is a regular repair task or a node operation task?

**Review by asias — APPROVED** [review]

**Comment by bhalevy** [comment]
@Deexie please stand me corrected if needed...

> It looks like the tasks listed here contains not only the higher level task but also the child tasks.

We should use the `internal` flag introduced in e2e8a286ccc7e23df4bdc95c79825a602e8593b2 to filter out child tasks in the tasks list by default.

> How long will the "done" tasks stay? When will we get rid of the tasks that have finished?

That the task_manager ttl configuration option that is 10 seconds by default.

> What is the sequence number? How is it numbered?

It's equivalent to the original repair ID and actually replaces it in the repair api.
It's incremented sequentially by each task manager module for each new top level task (a task that has no parent).

> What is entity?

Each module can use it any way it wants.
For example, the compaction manager may choose to store the compaction_group UUID there as an entity within the keyspace/table.

Repair could use it to represent the particular token range for the task.

> Does the progress actually work now?

I think it should in this PR, right, @Deexie?

**Comment by denesb** [comment]
I see all reviewers approved yet there is still ongoing discussion. Its not clear to me if the PR is good to go in and the discussion is about possible follow-ups, or there will be more changes coming.

**Review by bhalevy — APPROVED** [review]
For me it's good to go.
There are some follow ups regarding the visibility of internal repair tasks and potentially beefing up the per-task metadata we present over the API in list_tasks, but those can be done later.

**Comment by asias** [comment]
> @Deexie please stand me corrected if needed...
> 
> > It looks like the tasks listed here contains not only the higher level task but also the child tasks.
> 
> We should use the `internal` flag introduced in [e2e8a28](https://github.com/scylladb/scylladb/commit/e2e8a286ccc7e23df4bdc95c79825a602e8593b2) to filter out child tasks in the tasks list by default.

Let's use it then. The internal tasks is not very useful for normal users. It is enough to show to the high level tasks started by user themself. 

On the other hand, if the internal tasks is not  visible to the users, how is the internal task going to help / be useful for the users?

When we query the detailed info of a high level task, will it list the sub tasks?

> 
> > How long will the "done" tasks stay? When will we get rid of the tasks that have finished?
> 
> That the task_manager ttl configuration option that is 10 seconds by default.

This requires the user to query the task info in 10s when the task is done. It is possible that on some per-shard-task finish faster than the other tasks. So when the high level task is done, the user goes to check the task status, the tasks finished early have finished are gone already. 

> 
> > What is the sequence number? How is it numbered?
> 
> It's equivalent to the original repair ID and actually replaces it in the repair api. It's incremented sequentially by each task manager module for each new top level task (a task that has no parent).

So, all the child tasks share the same sequence number of the highest level task.

> 
> > What is entity?
> 
> Each module can use it any way it wants. For example, the compaction manager may choose to store the compaction_group UUID there as an entity within the keyspace/table.
> 
> Repair could use it to represent the particular token range for the task.

Entity is a string field? I saw it is empty. I suggest not to show it if it is empty.

Perhaps task_specific_info is a better name than entity. 

> 
> > Does the progress actually work now?
> 
> I think it should in this PR, right, @Deexie?

The progress shows it is zero.

````
"progress_units": "ranges",
"progress_total": 0,
"progress_completed": 0
```

**Comment by mykaul** [comment]
> For me it's good to go. There are some follow ups regarding the visibility of internal repair tasks and potentially beefing up the per-task metadata we present over the API in list_tasks, but those can be done later.

I'd be happy to see (in a follow-up) a documentation for the API and specifically, the entities in each API call.

**Comment by denesb** [comment]
Ok, queued then.

**Comment by bhalevy** [comment]
> > @Deexie please stand me corrected if needed...
> > > It looks like the tasks listed here contains not only the higher level task but also the child tasks.
> > 
> > 
> > We should use the `internal` flag introduced in [e2e8a28](https://github.com/scylladb/scylladb/commit/e2e8a286ccc7e23df4bdc95c79825a602e8593b2) to filter out child tasks in the tasks list by default.
> 
> Let's use it then. The internal tasks is not very useful for normal users. It is enough to show to the high level tasks started by user themself.
> 
> On the other hand, if the internal tasks is not visible to the users, how is the internal task going to help / be useful for the users?
> 

Normally the user need not be aware of the internal tasks.

I think it could be useful for diagnostics if e.g. a task hangs and one wants to know the progress of all sub-tasks to zoom-in on the problem.

> When we query the detailed info of a high level task, will it list the sub tasks?
> 

Not at the moment.
Maybe we can return an array of sub-tasks IDs so the user can get their particular status.

> > > How long will the "done" tasks stay? When will we get rid of the tasks that have finished?
> > 
> > 
> > That the task_manager ttl configuration option that is 10 seconds by default.
> 
> This requires the user to query the task info in 10s when the task is done. It is possible that on some per-shard-task finish faster than the other tasks. So when the high level task is done, the user goes to check the task status, the tasks finished early have finished are gone already.
> 

The parent task actually keeps a shared ptr of the sub-task so its life time is extended.
It will then be destroyed along with the parent.

> > > What is the sequence number? How is it numbered?
> > 
> > 
> > It's equivalent to the original repair ID and actually replaces it in the repair api. It's incremented sequentially by each task manager module for each new top level task (a task that has no parent).
> 
> So, all the child tasks share the same sequence number of the highest level task.
> 

Yes.

> > > What is entity?
> > 
> > 
> > Each module can use it any way it wants. For example, the compaction manager may choose to store the compaction_group UUID there as an entity within the keyspace/table.
> > Repair could use it to represent the particular token range for the task.
> 
> Entity is a string field? I saw it is empty. I suggest not to show it if it is empty.
> 

I don't mind either way.
@amnonh / @nyh do you have an opinion about this from the REST API usage perspective?

> Perhaps task_specific_info is a better name than entity.
> 

Hmm, @tzach @annastuchlik @amnonh what do you think?
This is an admin visible API so it should be self-descriptive as possible.

Regarding documentation, we have https://github.com/scylladb/scylladb/blob/master/api/api-doc/task_manager.json
that should be enough (we don't have yet another doc for API).
There, is says https://github.com/scylladb/scylladb/blob/c5121cf27365abf5989ed17078f2edff5ac59c1b/api/api-doc/task_manager.json#L234

> > > Does the progress actually work now?
> > 
> > 
> > I think it should in this PR, right, @Deexie?
> 
> The progress shows it is zero.
> 
> ```
> "progress_units": "ranges",
> "progress_total": 0,
> "progress_completed": 0
> ```
> ```

Yeah, I'm not sure what version @Deexie used to get this excerpt.
We should have a test that check this.

**Comment by Deexie** [comment]
> It looks like the tasks listed here contains not only the higher level task but also the child tasks. How many levels do we actually have? E.g., the one created by user from api (highest level), then each keysapces / tables / ranges, each shard?

We have the main task, per shard tasks and there will be per range tasks in the future. I will set the `internal` flag for all child tasks.

>  How long will the "done" tasks stay? When will we get rid of the tasks that have finished?
> 
> > That the task_manager ttl configuration option that is 10 seconds by default.

Or when the task_status or wait_task is performed.

> Does the progress actually work now?

No, not yet. It's going to be introduced together with per range tasks, since we measure it with ranges.

> How do I know if the task is a regular repair task or a node operation task?

A `type` field contains a stringified `streaming::stream_reason` so you can see it there.

> On the other hand, if the internal tasks is not visible to the users, how is the internal task going to help / be useful for the users?

There is an option in `list_module_tasks` that allows to show all the tasks.

**Pull request merged**